When merging examples in allOf, merge entries by index, not concat

ottomata opened the pull request:
When merging schemas, we can assume that
The examples in each schema
can each be merged together, so that the
the merged example entries will validate
against the final merged schema.
By default, json-schema-merge-allof concats arrays
to merge them, and JSONSchema examples is an array.

We want each element in each to-be-merged examples
to be merged together, not concat-ed.

I.e. we want:
  merge(schemaA.examples[0], schemaB.examples[0])

NOT the default behavior:
  schemaA.examples.concat(schemaB.examples)

Before merging, we also reverse the candidate examples.
We want to assume that schemas later in the allOf list
have more complete and specific examples and want any
final conflicting example values to take precedence.

This will make it easier to compose complete examples
from included schemas, just like we allow with the schemas themselves.

Bug: T242454

ottomata committed:
When merging examples in allOf, merge entries by index, not concat

When merging schemas, we can assume that
The examples in each schema
can each be merged together, so that the
the merged example entries will validate
against the final merged schema.
By default, json-schema-merge-allof concats arrays
to merge them, and JSONSchema examples is an array.

We want each element in each to-be-merged examples
to be merged together, not concat-ed.

I.e. we want:
  merge(schemaA.examples[0], schemaB.examples[0])

NOT the default behavior:
  schemaA.examples.concat(schemaB.examples)

Before merging, we also reverse the candidate examples.
We want to assume that schemas later in the allOf list
have more complete and specific examples and want any
final conflicting example values to take precedence.

This will make it easier to compose complete examples
from included schemas, just like we allow with the schemas themselves.

Bug: T242454

Pchelolo commented:
This looks very magical. What happens if one schema has just one example, and the other one has 3? What happens if it's vice versa? 

I'm wondering if an alternative approach would be better to allow examples only on the top level of the schema, and deference them with specific examples from fragments. Something like

```yaml
# revision-create.yaml

examples:
 allOf:
   - $ref: /fragment/mediawiki/revision/common/1.0.0#examples[0]
   - rev_id: 10
     rev_len: 42
     ...
```

This would require a bit more boilerplate, but imho this is much more understandable

ottomata commented:
In meeting today we decided to let this patch sit for a while.  It might be too magical.

bearloga commented:
@linehan and I just ran into this issue while figuring out the process for factoring out fields into a new "base" schema.

1. Start with mobile_apps/ios/schema_name which has examples
2. Factor out fields + examples into mobile_apps/base/schema_name, a fully specified schema in its own right
3. Change mobile_apps/ios/schema_name to `$ref` mobile_apps/base/schema_name, possibly add new fields at this point
4. Create mobile_apps/android/schema_name which also `$ref`s mobile_apps/base/schema_name, possibly add Android-specific fields at this point
5. Manually materialize base first and **then** iOS/Android off-shoots
6. Error in `npm test` because examples inherited in the off-shoots from base have `$schema: mobile_apps/base/schema_name/1.0.0`

So…maybe an argument against requiring every schema to have at least one example.

ottomata commented:
I ran into a problem today because a schema was merged without examples.  I wanted to require that all schemas have examples as part of https://phabricator.wikimedia.org/T270134, but then this problem hit me again.  fragment schemas are also real schemas, so they should have examples too.  If they do, then without this patch their examples will be merged into the concrete schema incorrectly.

ottomata committed:
Add test to require examples

Bug: T270134

ottomata commented:
@Pchelolo 
>  What happens if one schema has just one example, and the other one has 3? What happens if it's vice versa?

I just tested, and everything is ok. The merging only happens for corresponding entries based on index in the examples arrary.  If schema A has 1 example and schema B has 2, only the first example in A will be merged with the first example in B, and the second example in B will be left as is.

This is better than the current situation where examples will be concatenated together!

Pchelolo commented:
> I just tested, and everything is ok. The merging only happens for corresponding entries based on index in the examples arrary. If schema A has 1 example and schema B has 2, only the first example in A will be merged with the first example in B, and the second example in B will be left as is

Would it be more correct if the result is a product of all examples? e.g. merging sub-schema with examples [a, b] with schema with examples [c, d] should produce 4 examples, ac, bc, ad, bd ? That will make all the resulting examples valid. Now if we have 2 examples in super-schema, and just one in the sub-schema, we will end up with an invalid second example in the super-schema, since it won't be merged, no?

ottomata commented:
> Would it be more correct if the result is a product of all examples

Hm, interesting idea.

> Now if we have 2 examples in super-schema, and just one in the sub-schema, we will end up with an invalid second example in the super-schema, since it won't be merged, no?

This is true, but I think ok.  The tests will fail since the example won't be valid, and the dev can fix the examples in either schema. But I agree this is weird.  If the dev wants to add extra examples, it seems like they should be able to without having to modify the sub schema.

In reality, I'd like to add an override for our `/fragment` schemas, and not force them to have examples at all.  But for this more general library...hmm.

Maybe you are right.  We could also add an exception: if the number of examples is equal, merge like this, otherwise do a product? Another idea is to limit the product to only use the first example in any sub schema, but all examples in the last schema, e.g. [a, b], [c, d] => ac, ad.  That would generate fewer final examples, but might be more confusing.  

@linehan @bearloga @mdholloway, what do you think?

ottomata commented:
Also @marcelrf

bearloga commented:
> In reality, I'd like to add an override for our `/fragment` schemas, and not force them to have examples at all. But for this more general library...hmm.

It seems like there should be a way to disable both example inheritance and the example requirement.

**Re: product of examples.** What about extending the shorter list with repeating/recycling? In NumPy there's the concept of [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) and in R there's

> Vectors can be used in arithmetic expressions, in which case the operations are performed element by element. Vectors occurring in the same expression need not all be of the same length. If they are not, the value of the expression is a vector with the same length as the longest vector which occurs in the expression. Shorter vectors in the expression are recycled as often as need be (perhaps fractionally) until they match the length of the longest vector.

So:
- [a, b], [c, d] => [ab, cd]
- [a, b], [c] => [ac, bc]
- [a, b], [c, d, e] => [ac, bd, ae]

ottomata committed:
When merging examples, use cartesian product

This handles the case where the number of examples in each schema to merge
is different.

ottomata commented:
> Would it be more correct if the result is a product of all examples

@Pchelolo Am convinced.  Implemented.

> Re: product of examples. What about extending the shorter list with repeating/recycling?
@bearloga I like this idea too.  I took a quick stab at implementing that, but along the way I ended up liking the product better.  The cartesian product is probably more 'correct', but the 'broadcast' or 'limited product' is prettier.  Both are functional.  @Pchelolo got a preference one way or the other?  If you like Mikhael's idea I'll follow up.  Otherwise let me know if what I have with your product idea is good.

ottomata committed:
Default requireExamples: false

This should default to false for backwards compatibility.

ottomata committed:
Use lodash flatMap and flat; Node 10 doesn't have Array.flatMap or flat

Pchelolo commented:
>  @Pchelolo got a preference one way or the other? If you like Mikhael's idea I'll follow up. Otherwise let me know if what I have with your product idea is good.

As I understand 'broadcasting' will have the benefit of producing less examples in the end. After all, if we have a large number of nested schemas with large number of examples, we will end up multiplying all numbers of examples with simple product while with Mikhael's approach we will still be limited to the larges number of examples. This is all however purely theoretical and I like the simple product more simply cause I know what it is :) No strong preference of either.

Pchelolo commented on lib/jsonschema-tools.js:
```diff
@@ -515,6 +522,19 @@ function schemaPathToInfo(schemaPath, options = {}) {
+ * @return {Array}
+ */
+function cartesianProduct(...arrays) {
+    return arrays.reduce((a, b) => _.flatMap(a, d => b.map(e => [d, e].flat())));
```
[offtopic] This is SO unreadable, I'm impressed :) Nice that you added a comment about where it's taken from.

ottomata replied:
Hahah, I started read-ifying but gave up because it was taking to long.

Pchelolo reviewed (approved):
LGTM!

ottomata commented:
I'm going to remove the `requireExamples` bit from this patch and possibly add it in another.  Jason had an interesting idea of auto-generating examples from the schema, possibly using https://github.com/json-schema-faker/json-schema-faker.

ottomata committed:
Remove requireExamples

We may add this back in a later patch, or possibly
add support for auto generating examples instead.
